### PR TITLE
[Enhancement] Support recover partition version to latest tablet version when some metadata is lost

### PR DESCRIPTION
**Pull request by gengjun-git**
Why I'm doing:
When only FE metadata is lost, the query will report a version mismatch error, and the system needs to be able to recover the partition version based on the tablet version reported by BE.

What I'm doing:
Set the configuration `metadata_enable_recovery_mode` to true to recover the partition version based on the tablet version reported by BE.

Introduce a new syntax to manually set the version of a partition
```
admin set table table_name partition(partiiton_name/partition_id) to version xxx;
```

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [x] This pr needs user documentation (for new or modified features or behaviors)
  - [x] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/consistency/PartitionVersionRecovery.java [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+
+                        PARTITION:
+                        for (PhysicalPartition physicalPartition : partition.getSubPartitions()) {
+                            long recoverVersion = -1L;
+                            for (MaterializedIndex idx : physicalPartition.getMaterializedIndices(
+                                    MaterializedIndex.IndexExtState.VISIBLE)) {
+                                for (Tablet tablet : idx.getTablets()) {
+                                    LocalTablet localTablet = (LocalTablet) tablet;
+                                    long quorumVersion = localTablet.getQuorumVersion(replicaNum / 2 + 1);
+                                    if (quorumVersion == -1L) {
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/consistency/PartitionVersionRecovery.java:89]: As discussed, we need to be able to recover in this situation.

File: fe/fe-core/src/main/java/com/starrocks/consistency/PartitionVersionRecovery.java [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+import com.starrocks.persist.PartitionVersionRecoveryInfo;
+import com.starrocks.persist.PartitionVersionRecoveryInfo.PartitionVersion;
+import com.starrocks.server.GlobalStateMgr;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class PartitionVersionRecovery extends FrontendDaemon {
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/consistency/PartitionVersionRecovery.java:37]: As discussed, we need to explicitly tell the user when we have recovered all the tables, i.e. the partition visible version = quorum_of(replicas.version  of each tablet). For tablet which has been required for a long time, we can log the tablet id, just as decommission, for debugging purpose.

File: fe/fe-core/src/main/java/com/starrocks/transaction/GlobalTransactionMgr.java [hunk 3]
```diff
@@ -142,12 +144,18 @@ public long beginTransaction(long dbId, List<Long> tableIdList, String label, TU
...
+        if (Config.metadata_enable_recovery_mode) {
+            ErrorReportException.report(ErrorCode.ERR_BEGIN_TXN_FAILED,
+                    "The cluster is under recovery mode, all load jobs are rejected");
         }
 
         if (GlobalStateMgr.getCurrentState().isSafeMode()) {
-            throw new AnalysisException(String.format("The cluster is under safe mode state," +
-                    " all load jobs are rejected."));
+            ErrorReportException.report(ErrorCode.ERR_BEGIN_TXN_FAILED,
+                    "The cluster is under safe mode state, all load jobs are rejected.");
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/transaction/GlobalTransactionMgr.java:158]:
```suggestion
                    "The cluster is under recovery mode, all load jobs are rejected.");
```

to align with the configuration

File: fe/fe-core/src/main/java/com/starrocks/consistency/PartitionVersionRecovery.java [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+    private static final Logger LOG = LogManager.getLogger(PartitionVersionRecovery.class);
+
+    public PartitionVersionRecovery() {
+        super("partition_version_recovery");
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/consistency/PartitionVersionRecovery.java:41]:
```suggestion
        super("meta-recovery");
```

File: fe/fe-core/src/main/java/com/starrocks/consistency/PartitionVersionRecovery.java [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+                    OlapTable olapTable = (OlapTable) table;
+
+                    for (Partition partition : olapTable.getAllPartitions()) {
+                        short replicaNum = olapTable.getPartitionInfo().getReplicationNum(partition.getId());
+
+                        PARTITION:
+                        for (PhysicalPartition physicalPartition : partition.getSubPartitions()) {
+                            long recoverVersion = -1L;
+                            for (MaterializedIndex idx : physicalPartition.getMaterializedIndices(
+                                    MaterializedIndex.IndexExtState.VISIBLE)) {
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/consistency/PartitionVersionRecovery.java:85]: why only VISIBLE?

File: fe/fe-core/src/main/java/com/starrocks/consistency/PartitionVersionRecovery.java [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+                                        LOG.warn("cannot find quorum version for tablet: {}, " +
+                                                "ignore partition: {}-{} table: {}-{} db: {}-{}",
+                                                tablet.getId(), physicalPartition.getId(), partition.getName(),
+                                                table.getId(), table.getName(), database.getId(), database.getFullName());
+                                        continue PARTITION;
+                                    }
+                                    if (recoverVersion == -1L) {
+                                        recoverVersion = quorumVersion;
+                                    } else if (recoverVersion != quorumVersion) {
+                                        LOG.warn("found different quorum versions: {} {}, " +
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/consistency/PartitionVersionRecovery.java:99]: normally, this shouldn't happen, right?

File: fe/fe-core/src/main/java/com/starrocks/transaction/GlobalTransactionMgr.java [hunk 7]
```diff
@@ -142,12 +144,18 @@ public long beginTransaction(long dbId, List<Long> tableIdList, String label, TU
...
+
+        if (Config.metadata_enable_recovery_mode) {
+            ErrorReportException.report(ErrorCode.ERR_BEGIN_TXN_FAILED,
+                    "The cluster is under recovery mode, all load jobs are rejected");
         }
 
         if (GlobalStateMgr.getCurrentState().isSafeMode()) {
-            throw new AnalysisException(String.format("The cluster is under safe mode state," +
-                    " all load jobs are rejected."));
+            ErrorReportException.report(ErrorCode.ERR_BEGIN_TXN_FAILED,
```
[the remark this answers is not in the document]
Comment (gengjun-git) [reply fe/fe-core/src/main/java/com/starrocks/transaction/GlobalTransactionMgr.java:157]: Safemode is another feature introduced in https://github.com/StarRocks/starrocks/pull/21855. I just unify the error msg.

File: fe/fe-core/src/main/java/com/starrocks/system/SystemInfoService.java [hunk 8]
```diff
@@ -1247,6 +1248,17 @@ public void updatePathInfo(List<DiskInfo> addedDisks, List<DiskInfo> removedDisk
         LOG.debug("update path infos: {}", newPathInfos);
     }
 
+    public boolean checkTabletReportCacheUp(long timeMs) {
+        for (Backend backend : idToBackendRef.values()) {
+            if (TimeUtils.timeStringToLong(backend.getBackendStatus().lastSuccessReportTabletsTime) < timeMs) {
+                LOG.warn("last tablet report time of backend {}:{} is {}, should wait it to report tablets",
+                        backend.getHost(), backend.getHeartbeatErrMsg(), backend.getBackendStatus().lastSuccessReportTabletsTime);
```
[the remark this answers is not in the document]
Comment (gengjun-git) [reply fe/fe-core/src/main/java/com/starrocks/system/SystemInfoService.java:1198]: done

File: fe/fe-core/src/main/java/com/starrocks/transaction/GlobalTransactionMgr.java (hunk 3 above)
Comment (gengjun-git) [reply fe/fe-core/src/main/java/com/starrocks/transaction/GlobalTransactionMgr.java:158]: Safemode is another feature introduced in https://github.com/StarRocks/starrocks/pull/21855. I just unify the error msg.

File: fe/fe-core/src/main/java/com/starrocks/consistency/PartitionVersionRecovery.java (hunk 6 above)
Comment (gengjun-git) [reply fe/fe-core/src/main/java/com/starrocks/consistency/PartitionVersionRecovery.java]: Shadow index is an intermediate state for schema change. The tablets of shadow index maybe unhealthy at certain stages of schema change.

File: fe/fe-core/src/main/java/com/starrocks/consistency/PartitionVersionRecovery.java (hunk 5 above)
Comment (gengjun-git) [reply fe/fe-core/src/main/java/com/starrocks/consistency/PartitionVersionRecovery.java]: Shadow index is an intermediate state for schema change. The tablets of shadow index maybe unhealthy at certain stages of schema change.

File: fe/fe-core/src/main/java/com/starrocks/consistency/PartitionVersionRecovery.java (hunk 1 above)
Comment (gengjun-git) [reply fe/fe-core/src/main/java/com/starrocks/consistency/PartitionVersionRecovery.java]: done

File: fe/fe-core/src/main/java/com/starrocks/consistency/PartitionVersionRecovery.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (gengjun-git) [reply fe/fe-core/src/main/java/com/starrocks/consistency/PartitionVersionRecovery.java]: done

Comment (gengjun-git) [reply fe/fe-core/src/main/java/com/starrocks/consistency/PartitionVersionRecovery.java]: done

File: fe/fe-core/src/main/java/com/starrocks/system/SystemInfoService.java [hunk 9]
```diff
@@ -1247,6 +1248,17 @@ public void updatePathInfo(List<DiskInfo> addedDisks, List<DiskInfo> removedDisk
         LOG.debug("update path infos: {}", newPathInfos);
     }
 
+    public boolean checkTabletReportCacheUp(long timeMs) {
```
[the remark this answers is not in the document]
Comment (gengjun-git) [reply fe/fe-core/src/main/java/com/starrocks/system/SystemInfoService.java]: done

File: fe/fe-core/src/main/java/com/starrocks/server/GlobalStateMgr.java [hunk 10]
```diff
@@ -1816,6 +1829,11 @@ protected boolean replayJournalInner(JournalCursor cursor, boolean flowControl)
     }
 
     protected boolean canSkipBadReplayedJournal(Throwable t) {
+        if (Config.metadata_enable_recovery_mode) {
+            LOG.error("skip journal load failure because cluster is in recovery mode");
```
[the remark this answers is not in the document]
Comment (gengjun-git) [reply fe/fe-core/src/main/java/com/starrocks/server/GlobalStateMgr.java]: done

File: fe/fe-core/src/main/java/com/starrocks/common/ErrorCode.java [hunk 11]
```diff
@@ -332,7 +271,13 @@ public enum ErrorCode {
      * 5500 - 5599: DDL operation failure
      */
     ERR_LOC_AWARE_UNSUPPORTED_FOR_COLOCATE_TBL(5500, new byte[] {'4', '2', '0', '0', '0'},
-            "table '%s' has location property and cannot be colocated")
+            "table '%s' has location property and cannot be colocated"),
+    ERR_BAD_DB_ERROR(5501, new byte[] {'4', '2', '0', '0', '0'}, "Unknown database '%s'"),
+    ERR_BAD_TABLE_ERROR(5502, new byte[] {'4', '2', '0', '0', '0'}, "Unknown table '%s'"),
+    ERR_NOT_OLAP_TABLE(5503, new byte[] {'4', '2', '0', '0', '0'}, "Table '%s' is not a OLAP table"),
+    ERR_MULTI_SUB_PARTITION(5504, new byte[] {'4', '2', '0', '0', '0'},
+            "Partition '%s' has sub partitions, should specify the partition id"),
```
Comment (jaogoy) [inline fe/fe-core/src/main/java/com/starrocks/common/ErrorCode.java:279]: Sub-partition is not a user seeable concept. will it be returned to users?

File: fe/fe-core/src/main/java/com/starrocks/common/ErrorCode.java [hunk 12]
```diff
@@ -332,7 +271,13 @@ public enum ErrorCode {
      * 5500 - 5599: DDL operation failure
      */
     ERR_LOC_AWARE_UNSUPPORTED_FOR_COLOCATE_TBL(5500, new byte[] {'4', '2', '0', '0', '0'},
-            "table '%s' has location property and cannot be colocated")
+            "table '%s' has location property and cannot be colocated"),
+    ERR_BAD_DB_ERROR(5501, new byte[] {'4', '2', '0', '0', '0'}, "Unknown database '%s'"),
```
Comment (jaogoy) [inline fe/fe-core/src/main/java/com/starrocks/common/ErrorCode.java:275]: maybe 3F000 is better?

File: fe/fe-core/src/main/java/com/starrocks/common/ErrorCode.java [hunk 13]
```diff
@@ -332,7 +271,13 @@ public enum ErrorCode {
...
+    ERR_BAD_TABLE_ERROR(5502, new byte[] {'4', '2', '0', '0', '0'}, "Unknown table '%s'"),
```
Comment (jaogoy) [inline fe/fe-core/src/main/java/com/starrocks/common/ErrorCode.java:276]: 42602 may be better?

**Review by jaogoy — COMMENTED** [review]
Just for error code/SQLState

File: fe/fe-core/src/main/java/com/starrocks/common/ErrorCode.java (hunk 11 above)
Comment (gengjun-git) [reply fe/fe-core/src/main/java/com/starrocks/common/ErrorCode.java:279]: Yes, when setting the partition version of random distribution table, the partition id should be specified.

File: fe/fe-core/src/main/java/com/starrocks/common/ErrorCode.java (hunk 12 above)
Comment (gengjun-git) [reply fe/fe-core/src/main/java/com/starrocks/common/ErrorCode.java]: done

File: fe/fe-core/src/main/java/com/starrocks/common/ErrorCode.java (hunk 13 above)
Comment (gengjun-git) [reply fe/fe-core/src/main/java/com/starrocks/common/ErrorCode.java]: done

File: fe/fe-core/src/main/java/com/starrocks/consistency/MetaRecoveryDaemon.java [hunk 14]
```diff
@@ -0,0 +1,347 @@
...
+                    }
+
+                    OlapTable olapTable = (OlapTable) table;
+
+                    for (Partition partition : olapTable.getAllPartitions()) {
+                        short replicaNum = olapTable.getPartitionInfo().getReplicationNum(partition.getId());
+
+                        for (PhysicalPartition physicalPartition : partition.getSubPartitions()) {
+                            long recoverVersion = -1L;
+                            Map<Long, Boolean> commonVersion = new HashMap<>();
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/consistency/MetaRecoveryDaemon.java:96]: add comments for this var?

File: fe/fe-core/src/main/java/com/starrocks/consistency/MetaRecoveryDaemon.java [hunk 15]
```diff
@@ -0,0 +1,347 @@
...
+                                        LOG.warn("found different quorum versions: {} {}, " +
+                                                        "ignore partition: {}-{} table: {}-{} db: {}-{}",
+                                                recoverVersion, quorumVersion, physicalPartition.getId(), partition.getName(),
+                                                table.getId(), table.getName(), database.getId(), database.getFullName());
+                                    }
+
+                                    Set<Long> replicaVersions = localTablet.getAllReplicaVersions();
+                                    if (isFirstTablet) {
+                                        for (Long version : replicaVersions) {
+                                            commonVersion.put(version, true);
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/consistency/MetaRecoveryDaemon.java:129]: if `foundCommonVersion = true` , we don't need to update `commonVersion` map?

File: fe/fe-core/src/main/java/com/starrocks/consistency/MetaRecoveryDaemon.java [hunk 16]
```diff
@@ -0,0 +1,347 @@
...
+        }
+    }
+
+    public void fetchProcNodeResult(BaseProcResult result) {
+        this.lock.readLock().lock();
+        try {
+            for (UnRecoveredPartition info : unRecoveredPartitions) {
+                String advise;
+                if (info.getMaxCommonVersion() != -1L) {
+                    advise = String.format("You can set the partition version to %d, using command: " +
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/consistency/MetaRecoveryDaemon.java:258]: after `admin set max version`, can the tablet recover to ingestable state automatically?

File: fe/fe-core/src/main/java/com/starrocks/consistency/MetaRecoveryDaemon.java [hunk 17]
```diff
@@ -0,0 +1,347 @@
...
+                                    addToUnRecoveredPartitions(recoveryInfo);
+                                } else {
+                                    partitionsToRecover.add(new PartitionVersion(dbId, table.getId(),
+                                            physicalPartition.getId(), recoverVersion));
+                                }
+                            } else if (!foundCommonVersion) {
+                                long maxCommonVersion = -1L;
+                                for (Map.Entry<Long, Boolean> entry : commonVersion.entrySet()) {
+                                    if (entry.getValue() && entry.getKey() > maxCommonVersion) {
+                                        maxCommonVersion = entry.getKey();
```
Comment (nshangyiming) [inline fe/fe-core/src/main/java/com/starrocks/consistency/MetaRecoveryDaemon.java:160]: what if it doesn't have common version? like 2 bucket, replication_num = 3,  tablet1(2,2,2), tablet2(3,3,4)

Comment (gengjun-git) [reply fe/fe-core/src/main/java/com/starrocks/consistency/MetaRecoveryDaemon.java:160]: Not covered this case

File: fe/fe-core/src/main/java/com/starrocks/transaction/DatabaseTransactionMgr.java [hunk 18]
```diff
@@ -767,6 +767,30 @@ public List<TransactionState> getCommittedTxnList() {
         }
     }
 
+    public boolean hasCommittedTxnOnPartition(long tableId, long partitionId) {
```
[the remark this answers is not in the document]
Comment (gengjun-git) [reply fe/fe-core/src/main/java/com/starrocks/transaction/DatabaseTransactionMgr.java:771]: done

File: fe/fe-core/src/main/java/com/starrocks/consistency/MetaRecoveryDaemon.java (hunk 14 above)
Comment (gengjun-git) [reply fe/fe-core/src/main/java/com/starrocks/consistency/MetaRecoveryDaemon.java:96]: done

**Review by Seaven — APPROVED** [review]

**Review by nshangyiming — APPROVED** [review]

**Review by HangyuanLiu — APPROVED** [review]

**Review by jaogoy — APPROVED** [review]

**Pull request merged**